src/Server/TCPHandler.cpp

```diff
@@ -1285,37 +1267,57 @@ void TCPHandler::receiveQuery()
 
     readStringBinary(state.query, *in);
 
+    /// TODO unify interserver authentication (currently this code looks like a backdoor at a first glance)
```

Comment:
Why do you think it's like a backdoor?

Reply:
Because the code around "interserver secret mode" is a bit like "spaghetti" (see also reply on prev comment), there's a lack of logging, a lack of comments, a lack of assertions and some parts look really bad (e.g. `USER_INTERSERVER_MARKER`). And it allows to execute arbitrary query as arbitrary user. That's why I just don't trust this code :)

But after taking a deeper look I realized that it seems to work correctly (but it took some time to understand).

Reply:
Will remove the part about backdoor. However, I've just realized that it's not clear why `salt` is needed and why it must not be empty. It does not protect from replay attack, does not make bruteforce harder and seems like it's still possible to check if the secret string equals to some string from a dictionary. Am I missing something?

Reply:
Salt is supposed to protect from replay attack, so it mustn't be empty

Reply:
Yes, salt will protect from replay attack if server asks client to calculate hash with specific random salt, but seems like in current implementation salt is useless, because it's controlled by client (as well as `cluster`, `query`, `query_id` and `initial_user`)